Docs + some refactoring for MetaHasTraitsObject

jvkersch opened the pull request:
closes https://github.com/enthought/traits/issues/341

This arose out of trying to write a docstring for `MetaHasTraits`. When we tried to figure out during the ETS sprint what that class does, we saw that `MetaHasTraitsObject` is called only for the side effects in its `__init__` method. As a result, it made sense to replace that class by a good-to-honest function call. The bulk of the diff in this PR is doing precisely that.

I also started writing some tests; they probably cover about 10% of what `MetaHasTraits` does...

mdickinson commented on traits/has_traits.py:
```diff
@@ -416,14 +416,24 @@ def _property_method ( class_dict, name ):
 
     def __new__ ( cls, class_name, bases, class_dict ):
-        mhto = MetaHasTraitsObject( cls, class_name, bases, class_dict, False )
+        traits_meta_dict = create_traits_meta_dict(
```
It would be nice to pass the `is_category` argument by name, for readability.

mdickinson commented on traits/has_traits.py:
```diff
@@ -456,341 +466,322 @@ def remove_listener ( cls, listener, class_name = '' ):
 
-            elif isinstance( value, property ):
-                class_traits[ name ] = generic_trait
+    .. warning:: this function alters the `class_dict` input dictionary.
```
Would it be possible to add descriptions of the inputs and return value, using the usual NumPy-like docstring style?

mdickinson commented on traits/has_traits.py:
```diff
@@ -416,14 +416,24 @@ def _property_method ( class_dict, name ):
-        mhto = MetaHasTraitsObject( cls, class_name, bases, class_dict, False )
+        traits_meta_dict = create_traits_meta_dict(
+            class_name, bases, class_dict, False )
+        class_dict.update(traits_meta_dict)
```
These two lines are kinda weird: we're using `create_traits_meta_dict` both to mutate its input (the `class_dict`), *and* to return a new dict that we immediately merge with the `class_dict`. It would feel cleaner either to return a brand new dictionary that can be used immediately as our new class dict, or to mutate class_dict and return nothing.

I realise that we're hampered by the fact that this function is being used by the category machinery, too. I wonder whether it's worth having a clean version of this function that's just used by `MetaHasTraits`.

jvkersch replied:
> this function is being used by the category machinery

Yes, that's exactly why I took that tortuous route. I agree with having a clean function for `MetaHasTraits`, especially since the `Category` code seems on its way out.

mdickinson replied:
> especially since the Category code seems on its way out

We can only hope ...

jvkersch commented:
@mdickinson I think I addressed your comments and made it so that `create_traits_meta_dict` returns a new dictionary to be used as the class_dict, as you suggested.  In the `Category.__new__` method I also moved the constructs for the case with 2 base classes closer together, and I think this made it more readable (for this case, we still have to do things like populate a new dictionary only to pull it apart again).

jvkersch commented:
I changed my mind on one thing: given that we don't need the original class_dict for anything (and that having it around is positively confusing) I changed the helper method so that it just modifies the dictionary in place and doesn't return anything.

mdickinson commented:
> I changed the helper method so that it just modifies the dictionary in place and doesn't return anything.

Sounds reasonable. Will review anon.

mdickinson commented:
Merged master into this branch, fixed conflicts and tests. The conflict fixing wasn't trivial; this will need a thorough re-review.

mdickinson commented on traits/category.py:
```diff
@@ -64,28 +78,6 @@ def __new__ ( cls, class_name, bases, class_dict ):
-#  'MetaCategoryObject' class:
-#-------------------------------------------------------------------------------
-
-class MetaCategoryObject ( MetaHasTraitsObject ):
```
There are still references to `MetaCategoryObject` and `MetaHasTraitsObject` in the documentation (`/docs/source/traits_api_reference/category.rst`).

mdickinson commented on traits/has_traits.py:
```diff
@@ -258,7 +258,7 @@ def _get_instance_handlers ( class_dict, bases ):
                         arg_lists.append( arg_list )
 
     # Merge in the information from the class dictionary:
-    for name, value in class_dict.items():
```
This change could probably be reverted (I may have accidentally introduced it during the merge).

mdickinson replied:
Done.

mdickinson commented on traits/has_traits.py:
```diff
@@ -439,344 +448,338 @@ def remove_listener ( cls, listener, class_name = '' ):
-                    prefix_traits[ name ] = value
+    This is called during the construction of a new HasTraits class. The first
+    three parameters have the same interpretation as the corresponding
+    parameters of ``type.__new__``. This function modifies `class_dict`
```
Nit: use double backticks around "class_dict"

mdickinson replied:
Fixed.

mdickinson commented on traits/has_traits.py:
```diff
@@ -439,344 +448,338 @@ def remove_listener ( cls, listener, class_name = '' ):
+                    class_traits[ name ] = value = ictrait( default_value )
+                    # Make sure that the trait now has the default value
+                    # has the correct initializer.
+                    value.default_value(1, value.default)
```
`1` should be replaced with `MISSING_DEFAULT_VALUE` here

mdickinson replied:
Done.

jvkersch commented:
@mdickinson This looks good; thanks for salvaging this PR and bringing it up-to-date!

What should the next step be? Nuking Categories?

mdickinson commented:
> Nuking Categories?

Yes, please!
